Make abomination stylesheet blog post (#1791)
bring/developer-site: content/blog/abomination-stylesheet.md

---
layout: post
title: Abomination stylesheet
excerpt: A low-level HTML markup quality assurance tool
publishDate: 2024-11-27 06:35:00
tags: 
    - CSS
authors:
    - Igor-J86
---

## What is "abomination"?
Once in a while you come across some words or phrases that you haven‘t heard before. The term "abomination" was totally strange to me, and after looking it up, it was surprising that the meaning of it is:

***Something that you dislike, disapprove, or something that is regarded with disgust.***

Huh... Should the thing we create show something that is disliked, disapproved or disgusting? Why?
The people we work with are good and smart developers, so why should we create something that can be destructive to other developer‘s UI implementations?

> It all became clear when I realised that the intention of such a thing is not to destroy or make things uglier, but rather making certain elements in the UI "stand out" in some way, signaling that there‘s an issue there.

Reviewer: I don’t think you need quotes around stand out, it’s a common term. If you want stand out to stand out, like applying pressure to it, you can make it italic.

### What is it?

> It is just vanilla CSS, where we utilise the power of CSS wildcards, pseudo-class functions and nesting. Using these techniques we can make certain CSS definitions based on semantic HTML markup, and address the issues that arise with invalid HTML markup by making those elements "stand out" and become eye catching.

Reviewer: Also "stand out" again, not wrong to repeat it, but maybe it can be phrased differently.

```html
<!-- ❌ DON‘T -->
<p>
  <div>Doing something smart here</div>
  Some text here...
</p>

<!-- ❌ DON‘T -->
<button>
  <a href="#">Label</a>
</button>

<!-- ❌ DON‘T -->
<a href="#">
  <button>Label</button>
</a>

<!-- ❌ DON‘T -->
<ul>
  <div>Doing something smart here</div>
  <li>Item 1</li>
  <li>Item 2</li>
  <li>Item 3</li>
</ul>

<!-- ❌ DON‘T -->
<label>Input label</label>
<input type="text" placeholder="user@example.com">
```

### How does it work?

> We can start by selecting the elements we want to target, where we have seen developers some times tend to write invalid HTML markup, either by mistake or trying to make something really clever. Some of the common mistakes that can be found around are `<div>` element inside `<p>`  elements, a `<button>` element containing an `<a>` element (or the other way around), and the one and only; `<label>` missing the `for` attribute or not wrapping the `<input>` element.

Reviewer: This holds a central tip about selecting typical mistakes and invalid HTML in our applications, and not making something universal to catch everything. That’s very valuable.

> We can start by selecting the elements we want to target, where we have seen developers some times tend to write invalid HTML markup, either by mistake or trying to make something really clever. Some of the common mistakes that can be found around are `<div>` element inside `<p>`  elements, a `<button>` element containing an `<a>` element (or the other way around), and the one and only; `<label>` missing the `for` attribute or not wrapping the `<input>` element.

Reviewer: Oh, and … a general writing tip on not coming across as too harsh, which I get the impression you are attempting here:
```We can start by selecting the elements we want to target, where we have seen developers some times tend to write invalid HTML markup```

This makes a distinction between we/us and developers, and can be seen as assigning blame or pointing finger at someone that is not us, and so it’s modified by adding "sometimes tend to" etc. Maybe "sometimes" and "tend to" cancel each other out too, but my point is, such modifiers can be cut if we use "we" consistently. We are making the mistakes, we are targeting the ones we tend to make, we are fixing them – it sounds more inclusive, it’s easier for people to accept that they do it when it’s all of us, which it is.

---
layout: post
title: Abomination stylesheet
excerpt: A low-level HTML markup quality assurance tool
publishDate: 2024-11-27 06:35:00
tags: 
    - CSS
authors:
    - Igor-J86
---

> The different browsers today are pretty forgiving when it comes to HTML markup and CSS. They have become good at rendering the markup, even if it‘s invalid HTML markup and will cause HTML validation to fail. While we should always strive to write valid HTML markup, mistakes do happen, and luckily the browsers do a pretty good job at rendering the code without breaking the layout completely. But, we still have invalid HTML markup, and our goal was to address these issues by making a low-level HTML markup quality assurance tool with CSS.

Reviewer: Any reason to write "HTML markup" and not just "HTML"?
Can they be mixed to create variation?

> The intention of such a tool is that we suddenly have continuous live testing tool for markup code across different teams / applications, with no need of additional dependencies. It can be included in the `<head>` of the project as a standalone stylesheet, based on the environment we want it to appear in, or import it into an existing CSS.

Reviewer: ```suggestion
The intention of such a tool is to continuously live test markup code across different teams/applications, with no need of additional dependencies. It can be included in the `<head>` of the project as a standalone stylesheet, based on the environment we want it to appear in, or import it into an existing CSS.
```

It is maybe not enough to just make an element stand out, but we should also inform about what the issue is in some way. A nice thing that comes in handy then is a custom property we can set: `--note`

This will show up in the developer tools (console) when you inspect the element that is suddenly surrounded by an ugly red outline, and can quickly identify what the fuzz is all about.

But, wait... By the power of CSS-Gods; we can actually display the note directly in the UI as well! 🤯

The whole CSS can be nested inside a wildcard selector. This will target all the elements that are defined inside and already marked with a red outline. By hovering these, we can then also display the `--note` that is applied to each of them inside an `::after` element.

```css
* {
  /* nested elements here */
  
  &:hover::after {
	content: var(--note);
	background-color: white;
	border: 2px solid var(--red);
	padding: 1rem;
	position: fixed;
	top: 20rem;
	left: 50%;
	transform: translateX(-50%);
	box-shadow: var(--shadow-dark);
	z-index: 9999;
  }
}

/*
  If you don‘t want elements to be selected
  by the wildcard selector, just use a :not()
  on it: *:not(.some-element)
*/
```

## How we use it

> In Mybring, we are currently linking to the abomination stylesheet directly within the `<head>` tag, but based on the testing environments we have. We don‘t want this to be available in the production environment. 😅

Reviewer: ```suggestion
In Mybring, we are currently linking to the abomination stylesheet within the `<head>` tag in our testing environments. We don‘t want this to be available in the production environment. 😅
```
